File: leetcode/0054-spiral-matrix/solution.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int m=matrix.length;
        int n=matrix[0].length;
        int left=0;
        int top=0;
        int down=m-1;
        int right=n-1;
        ArrayList<Integer>list=new ArrayList<>();
        while(left <= right && top <= down){
            for(int i=left;i<=right;i++){
                list.add(matrix[top][i]);
            }
            top++;
            for(int i=top;i<=down;i++){
                list.add(matrix[i][right]);
            }
            right--;
            if (top <= down) {
                for (int i = right; i >= left; i--) {
                    list.add(matrix[down][i]);
                }
                down--;
            }
            if (left <= right) {
                for (int i = down; i >= top; i--) {
                    list.add(matrix[i][left]);
                }
                left++;
            }
        }
        return list;
    }
}

File: leetcode/0054-spiral-matrix/solution.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        n,m = len(matrix),len(matrix[0])
        print(n,m)
        left , right = 0,m-1
        top,down = 0,n-1
        arr = []
        while left<=right and top<=down:
            for i in range(left,right+1):
                arr.append(matrix[top][i])
            top+=1
            
            for i in range(top,down+1):
                arr.append(matrix[i][right])
            right -=1
            
            if top <=down:
                for i in range(right,left-1,-1):
                    arr.append(matrix[down][i])
                down -=1
            if left<=right:
                for i in range(down,top-1,-1):
                    arr.append(matrix[i][left])
                left+=1
        print(arr)
        return arr

        
        
